crates/lair_keystore_api/src/lair_server/priv_api.rs

```diff
@@ -233,6 +233,127 @@ pub(crate) fn priv_req_new_seed<'a>(
         Ok(())
     }
 }
+pub(crate) fn priv_req_derive_seed<'a>(
+    inner: &'a Arc<RwLock<SrvInner>>,
+    send: &'a crate::sodium_secretstream::S3Sender<LairApiEnum>,
+    dec_ctx_key: &'a sodoken::BufReadSized<32>,
+    unlocked: &'a Arc<atomic::AtomicBool>,
+    req: LairApiReqDeriveSeed,
+) -> impl Future<Output = LairResult<()>> + 'a + Send {
+    async move {
+        let store = priv_get_store(inner, unlocked)?;
+
+        let entry = store.get_entry_by_tag(req.src_tag.clone()).await?;
+
+        let (src_seed, src_seed_info, src_dlp) = match (&*entry, req.src_deep_lock_passphrase) {
+            (LairEntryInner::Seed { seed, seed_info, .. }, None) => {
+                let seed = seed.decrypt(store.get_bidi_ctx_key()).await?;
+                (seed, seed_info, None)
+            }
+            (
+                LairEntryInner::DeepLockedSeed {
+                    seed,
+                    seed_info,
+                    ops_limit,
+                    mem_limit,
+                    salt,
+                    ..
+                },
+                Some(passphrase),
+            ) => {
+                let dlp = DeepLockPassphrase {
+                    ops_limit: *ops_limit,
+                    mem_limit: *mem_limit,
+                    passphrase,
+                };
+                let deep_key = deep_unlock_key_from_passphrase(
+                    &dlp,
+                    dec_ctx_key,
+                    BufReadSized::from(salt.clone().cloned_inner()),
+                )
+                .await?;
+                let seed = seed.decrypt(deep_key).await?;
+                (seed, seed_info, Some(dlp))
+            },
+            (LairEntryInner::WkaTlsCert { .. }, _) => return Err("The tag provided is for a Cert, which cannot be derived. You must specify the tag for a Seed.".into()),
+            (LairEntryInner::Seed { .. }, Some(_)) => return Err("A passphrase was provided for a seed which is not deep-locked. Make the request without a `src_passphrase`.".into()),
+            (LairEntryInner::DeepLockedSeed { .. }, None) => return Err("A `src_passphrase` is needed to unlock the source seed which is deep-locked.".into()),
+        };
+
+        let mut parent = src_seed;
+
+        for index in req.derivation_path.iter() {
+            let derived = BufWriteSized::new_mem_locked()?;
+            sodoken::kdf::derive_from_key(
+                derived.clone(),
+                *index as u64,
+                *b"IdnSecKy",
+                parent,
+            )?;
+            parent = derived.clone().into();
+        }
+
+        let dst_seed = parent;
+
+        // TODO: do we want to take the limits from the source passphrase if possible?
+        // TODO: do we even want to allow a destination passphrase if there is no source passphrase?
+        let (ops_limit, mem_limit) = if let Some(dlp) = src_dlp {
+            (dlp.ops_limit, dlp.mem_limit)
+        } else {
+            let limits = PwHashLimits::current();
```

Comment:
This call must move up out of the `async move`. If there are any awaits above it, there's a chance the original limits will no longer be in context.

Reply:
aw, crud. it's even worse than that. I think you need to move it outside the `async move` in the main public function, and then pass it into here.

Reply:
Yeah, use new_seed as a template. The [limits are captured at the top](https://github.com/holochain/lair/blob/a4fa49a2b8dcebe5e784ed44d2c22d6044a8abdc/crates/lair_keystore_api/src/lair_client.rs#L189) and then [forwarded to the server](https://github.com/holochain/lair/blob/a4fa49a2b8dcebe5e784ed44d2c22d6044a8abdc/crates/lair_keystore_api/src/lair_client.rs#L206-L207) in the request.